In [1]:
import pandas as pd
import numpy as np

In [2]:
customers = pd.read_csv('/mnt/HC_Volume_18315164/home-jupyter/jupyter-b-argymbek-18/Lesson_1/olist_customers_dataset.csv', encoding='Windows-1251', sep = ",")

In [3]:
customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [4]:
order_items = pd.read_csv('/mnt/HC_Volume_18315164/home-jupyter/jupyter-b-argymbek-18/Lesson_1/olist_order_items_dataset.csv', encoding='Windows-1251', sep = ",")

In [5]:
order_items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [6]:
orders = pd.read_csv('/mnt/HC_Volume_18315164/home-jupyter/jupyter-b-argymbek-18/Lesson_1/olist_orders_dataset.csv', encoding='Windows-1251', sep = ",")

In [7]:
orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [8]:
cust_ord = customers.merge(orders, how='left', on='customer_id')

In [9]:
cust_ord 

customer_id                customer_unique_id  \
0      06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1      18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2      4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3      b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4      4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                 ...                               ...   
99436  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
99437  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
99438  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
99439  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
99440  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

       customer_zip_code_prefix          customer_city customer_state  \
0                         14409                 franca             SP   
1                          9790  sao bernardo do campo             SP   
2                          1151              sao paulo             SP   
3                          8775        mogi das cruzes             SP   
4                         13056               campinas             SP   
...                         ...                    ...            ...   
99436                      3937              sao paulo             SP   
99437                      6764        taboao da serra             SP   
99438                     60115              fortaleza             CE   
99439                     92120                 canoas             RS   
99440                      6703                  cotia             SP   

                               order_id order_status order_purchase_timestamp  \
0      00e7ee1b050b8499577073aeb2a297a1    delivered      2017-05-16 15:05:35   
1      29150127e6685892b6eab3eec79f59c7    delivered      2018-01-12 20:48:24   
2      b2059ed67ce144a36e2aa97d2c9e9ad2    delivered      2018-05-19 16:07:45   
3      951670f92359f4fe4a63112aa7306eba    delivered      2018-03-13 16:06:38   
4      6b7d50bd145f6fc7f33cebabd7e49d0f    delivered      2018-07-29 09:51:30   
...                                 ...          ...                      ...   
99436  6760e20addcf0121e9d58f2f1ff14298    delivered      2018-04-07 15:48:17   
99437  9ec0c8947d973db4f4e8dcf1fbfa8f1b    delivered      2018-04-04 08:20:22   
99438  fed4434add09a6f332ea398efd656a5c    delivered      2018-04-08 20:11:50   
99439  e31ec91cea1ecf97797787471f98a8c2    delivered      2017-11-03 21:08:33   
99440  28db69209a75e59f20ccbb5c36a20b90    delivered      2017-12-19 14:27:23   

         order_approved_at order_delivered_carrier_date  \
0      2017-05-16 15:22:12          2017-05-23 10:47:57   
1      2018-01-12 20:58:32          2018-01-15 17:14:59   
2      2018-05-20 16:19:10          2018-06-11 14:31:00   
3      2018-03-13 17:29:19          2018-03-27 23:22:42   
4      2018-07-29 10:10:09          2018-07-30 15:16:00   
...                    ...                          ...   
99436  2018-04-07 16:08:45          2018-04-11 02:08:36   
99437  2018-04-04 08:35:12          2018-04-05 18:42:35   
99438  2018-04-08 20:30:03          2018-04-09 17:52:17   
99439  2017-11-03 21:31:20          2017-11-06 18:24:41   
99440  2017-12-19 18:50:39          2017-12-21 19:17:21   

      order_delivered_customer_date order_estimated_delivery_date  
0               2017-05-25 10:35:35           2017-06-05 00:00:00  
1               2018-01-29 12:41:19           2018-02-06 00:00:00  
2               2018-06-14 17:58:51           2018-06-13 00:00:00  
3               2018-03-28 16:04:25           2018-04-10 00:00:00  
4               2018-08-09 20:55:48           2018-08-15 00:00:00  
...                             ...                           ...  
99436           2018-04-13 20:06:37           2018-04-25 00:00:00  


In [10]:
cust_ord_items = cust_ord.merge(order_items, how='left', on='order_id')

In [11]:
cust_ord_items

customer_id                customer_unique_id  \
0       06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1       18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2       4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3       b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4       4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                  ...                               ...   
113420  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
113421  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
113422  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
113423  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
113424  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

        customer_zip_code_prefix          customer_city customer_state  \
0                          14409                 franca             SP   
1                           9790  sao bernardo do campo             SP   
2                           1151              sao paulo             SP   
3                           8775        mogi das cruzes             SP   
4                          13056               campinas             SP   
...                          ...                    ...            ...   
113420                      3937              sao paulo             SP   
113421                      6764        taboao da serra             SP   
113422                     60115              fortaleza             CE   
113423                     92120                 canoas             RS   
113424                      6703                  cotia             SP   

                                order_id order_status  \
0       00e7ee1b050b8499577073aeb2a297a1    delivered   
1       29150127e6685892b6eab3eec79f59c7    delivered   
2       b2059ed67ce144a36e2aa97d2c9e9ad2    delivered   
3       951670f92359f4fe4a63112aa7306eba    delivered   
4       6b7d50bd145f6fc7f33cebabd7e49d0f    delivered   
...                                  ...          ...   
113420  6760e20addcf0121e9d58f2f1ff14298    delivered   
113421  9ec0c8947d973db4f4e8dcf1fbfa8f1b    delivered   
113422  fed4434add09a6f332ea398efd656a5c    delivered   
113423  e31ec91cea1ecf97797787471f98a8c2    delivered   
113424  28db69209a75e59f20ccbb5c36a20b90    delivered   

       order_purchase_timestamp    order_approved_at  \
0           2017-05-16 15:05:35  2017-05-16 15:22:12   
1           2018-01-12 20:48:24  2018-01-12 20:58:32   
2           2018-05-19 16:07:45  2018-05-20 16:19:10   
3           2018-03-13 16:06:38  2018-03-13 17:29:19   
4           2018-07-29 09:51:30  2018-07-29 10:10:09   
...                         ...                  ...   
113420      2018-04-07 15:48:17  2018-04-07 16:08:45   
113421      2018-04-04 08:20:22  2018-04-04 08:35:12   
113422      2018-04-08 20:11:50  2018-04-08 20:30:03   
113423      2017-11-03 21:08:33  2017-11-03 21:31:20   
113424      2017-12-19 14:27:23  2017-12-19 18:50:39   

       order_delivered_carrier_date order_delivered_customer_date  \
0               2017-05-23 10:47:57           2017-05-25 10:35:35   
1               2018-01-15 17:14:59           2018-01-29 12:41:19   
2               2018-06-11 14:31:00           2018-06-14 17:58:51   
3               2018-03-27 23:22:42           2018-03-28 16:04:25   
4               2018-07-30 15:16:00           2018-08-09 20:55:48   
...                             ...                           ...   
113420          2018-04-11 02:08:36           2018-04-13 20:06:37   
113421          2018-04-05 18:42:35           2018-04-11 18:54:45   
113422          2018-04-09 17:52:17           2018-05-09 19:03:15   
113423          2017-11-06 18:24:41           2017-11-16 19:58:39   
113424          2017-12-21 19:17:21           2017-12-26 18:42:36   

       order_est

In [12]:
cust_ord_items.dtypes

customer_id                       object
customer_unique_id                object
customer_zip_code_prefix           int64
customer_city                     object
customer_state                    object
order_id                          object
order_status                      object
order_purchase_timestamp          object
order_approved_at                 object
order_delivered_carrier_date      object
order_delivered_customer_date     object
order_estimated_delivery_date     object
order_item_id                    float64
product_id                        object
seller_id                         object
shipping_limit_date               object
price                            float64
freight_value                    float64
dtype: object

In [13]:
cust_ord_items['order_purchase_timestamp'] = pd.to_datetime(cust_ord_items['order_purchase_timestamp'], errors='coerce')

In [14]:
cust_ord_items['order_approved_at'] = pd.to_datetime(cust_ord_items['order_approved_at'], errors='coerce')

In [15]:
cust_ord_items['order_delivered_carrier_date'] = pd.to_datetime(cust_ord_items['order_delivered_carrier_date'], errors='coerce')

In [16]:
cust_ord_items['order_delivered_customer_date'] = pd.to_datetime(cust_ord_items['order_delivered_customer_date'], errors='coerce')

In [17]:
cust_ord_items['order_estimated_delivery_date'] = pd.to_datetime(cust_ord_items['order_estimated_delivery_date'], errors='coerce')

In [18]:
cust_ord_items['shipping_limit_date'] = pd.to_datetime(cust_ord_items['shipping_limit_date'], errors='coerce')

In [19]:
cust_ord_items.dtypes

customer_id                              object
customer_unique_id                       object
customer_zip_code_prefix                  int64
customer_city                            object
customer_state                           object
order_id                                 object
order_status                             object
order_purchase_timestamp         datetime64[ns]
order_approved_at                datetime64[ns]
order_delivered_carrier_date     datetime64[ns]
order_delivered_customer_date    datetime64[ns]
order_estimated_delivery_date    datetime64[ns]
order_item_id                           float64
product_id                               object
seller_id                                object
shipping_limit_date              datetime64[ns]
price                                   float64
freight_value                           float64
dtype: object

In [20]:
cust_ord_items.order_estimated_delivery_date

0        2017-06-05
1        2018-02-06
2        2018-06-13
3        2018-04-10
4        2018-08-15
            ...    
113420   2018-04-25
113421   2018-04-20
113422   2018-05-02
113423   2017-12-05
113424   2018-01-08
Name: order_estimated_delivery_date, Length: 113425, dtype: datetime64[ns]

In [21]:
cust_ord_items.shape

(113425, 18)

In [22]:
cust_ord_items.isna().sum()

customer_id                         0
customer_unique_id                  0
customer_zip_code_prefix            0
customer_city                       0
customer_state                      0
order_id                            0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 161
order_delivered_carrier_date     1968
order_delivered_customer_date    3229
order_estimated_delivery_date       0
order_item_id                     775
product_id                        775
seller_id                         775
shipping_limit_date               775
price                             775
freight_value                     775
dtype: int64

In [23]:
df = cust_ord_items.drop_duplicates()

In [24]:
df

customer_id                customer_unique_id  \
0       06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1       18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2       4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3       b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4       4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                  ...                               ...   
113420  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
113421  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
113422  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
113423  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
113424  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

        customer_zip_code_prefix          customer_city customer_state  \
0                          14409                 franca             SP   
1                           9790  sao bernardo do campo             SP   
2                           1151              sao paulo             SP   
3                           8775        mogi das cruzes             SP   
4                          13056               campinas             SP   
...                          ...                    ...            ...   
113420                      3937              sao paulo             SP   
113421                      6764        taboao da serra             SP   
113422                     60115              fortaleza             CE   
113423                     92120                 canoas             RS   
113424                      6703                  cotia             SP   

                                order_id order_status  \
0       00e7ee1b050b8499577073aeb2a297a1    delivered   
1       29150127e6685892b6eab3eec79f59c7    delivered   
2       b2059ed67ce144a36e2aa97d2c9e9ad2    delivered   
3       951670f92359f4fe4a63112aa7306eba    delivered   
4       6b7d50bd145f6fc7f33cebabd7e49d0f    delivered   
...                                  ...          ...   
113420  6760e20addcf0121e9d58f2f1ff14298    delivered   
113421  9ec0c8947d973db4f4e8dcf1fbfa8f1b    delivered   
113422  fed4434add09a6f332ea398efd656a5c    delivered   
113423  e31ec91cea1ecf97797787471f98a8c2    delivered   
113424  28db69209a75e59f20ccbb5c36a20b90    delivered   

       order_purchase_timestamp   order_approved_at  \
0           2017-05-16 15:05:35 2017-05-16 15:22:12   
1           2018-01-12 20:48:24 2018-01-12 20:58:32   
2           2018-05-19 16:07:45 2018-05-20 16:19:10   
3           2018-03-13 16:06:38 2018-03-13 17:29:19   
4           2018-07-29 09:51:30 2018-07-29 10:10:09   
...                         ...                 ...   
113420      2018-04-07 15:48:17 2018-04-07 16:08:45   
113421      2018-04-04 08:20:22 2018-04-04 08:35:12   
113422      2018-04-08 20:11:50 2018-04-08 20:30:03   
113423      2017-11-03 21:08:33 2017-11-03 21:31:20   
113424      2017-12-19 14:27:23 2017-12-19 18:50:39   

       order_delivered_carrier_date order_delivered_customer_date  \
0               2017-05-23 10:47:57           2017-05-25 10:35:35   
1               2018-01-15 17:14:59           2018-01-29 12:41:19   
2               2018-06-11 14:31:00           2018-06-14 17:58:51   
3               2018-03-27 23:22:42           2018-03-28 16:04:25   
4               2018-07-30 15:16:00           2018-08-09 20:55:48   
...                             ...                           ...   
113420          2018-04-11 02:08:36           2018-04-13 20:06:37   
113421          2018-04-05 18:42:35           2018-04-11 18:54:45   
113422          2018-04-09 17:52:17           2018-05-09 19:03:15   
113423          2017-11-06 18:24:41           2017-11-16 19:58:39   
113424          2017-12-21 19:17:21           2017-12-26 18:42:36   

       order_estimated_deliv

Задание 1. Сколько у нас пользователей, которые совершили покупку только один раз?

Для того чтобы понять сколько у нас пользователей, которые совершили покупку только один раз, надо решить что считать за покупку? 
В первом варианте я использовал  order_approved_at так как это  —  время подтверждения оплаты заказа, при order_status != unavailable и order_status != canceled, так как unavailable — недоступен, можно предположить, что нет в наличии на текущий момент, а canceled - отменен, можно предположить, что помещают в корзину, но не оплачивают. 
Во втором варианте я использовал order_id так как это - уникальный идентификатор заказа (номер чека), также при order_status != unavailable и order_status != canceled.

In [25]:
res_1 = df.groupby(['customer_unique_id', 'order_status', 'order_approved_at'], as_index=False)\
        .agg({'order_approved_at': 'count'})\
        .query("order_approved_at == 1 & order_status != 'unavailable' & order_status != 'canceled'")\
        .rename(columns={'ordorder_approved_ater_id':'quantity_of_orders'}).shape[0]

In [26]:
res_1

88296

In [27]:
res_2 = df.groupby(['customer_unique_id', 'order_status', 'order_id'], as_index=False)\
        .agg({'order_id': 'count'})\
        .query("order_id == 1 & order_status != 'unavailable' & order_status != 'canceled'")\
        .rename(columns={'order_id':'quantity_of_orders'}).shape[0]

In [28]:
# Склоняюсь к этому варианту ответа
res_2

88446

Задание 2. Сколько заказов в месяц в среднем не доставляется по разным причинам (вывести детализацию по причинам)?

1.Отфильтровал заказы со статусами canceled и unavailable, так как canceled - отменен, можно предположить, что помещают в корзину, но не оплачивают или передумавают, а unavailable — недоступен, можно предположить, что нет в наличии на текущий момент
2.Добавил месяц на основе order_purchase_timestamp
3.Рассчитал уникальное количество заказов, для каждого месяца и каждого выбранного статуса
4.Рассчитал среднее количество закзаов по месяцам для каждого статуса


In [29]:
# Отфильтруем заказы со статусами canceled и unavailable
df_1 = df.query("order_status in ['canceled','unavailable']")

In [30]:
df_1.head()

customer_id                customer_unique_id  \
50   f34a6e874087ec1f0e3dab9fdf659c5d  233896de79986082f1f479f1f85281cb   
330  5bfe800011656c0afb81db64519982db  0071f46a072a9ae25bbe4438b15efe9c   
422  b08064e24083fee8fbe8797902b07ecd  035f60af6e7d7f78470e9443be08d339   
427  3ba47b09546c2da91ef141efce2d5888  b8c37ee810b38c416022e1c9f51e740c   
429  cfffbb322333151f8df6398a8c3340db  0dc0731aafd5659410244092b0dc7d65   

     customer_zip_code_prefix customer_city customer_state  \
50                      38300     ituiutaba             MG   
330                     16011     aracatuba             SP   
422                     28495       aperibe             RJ   
427                     32140      contagem             MG   
429                      5093     sao paulo             SP   

                             order_id order_status order_purchase_timestamp  \
50   6e98de3a85c84ead6689189b825d35b5     canceled      2018-03-15 10:07:02   
330  df8c077268f7f3baaac0892eb3143642  unavailable      2017-02-01 00:04:17   
422  c609f82bcf7a90292a5940205ebd7e93  unavailable      2018-05-13 16:45:55   
427  a402af38a74c44926247db438245d95c     canceled      2018-01-30 22:12:10   
429  f63cd97f8f1db2eb266fb3bc2b4a95f3     canceled      2017-06-19 14:07:04   

      order_approved_at order_delivered_carrier_date  \
50  2018-03-15 10:29:33                          NaT   
330 2017-02-01 00:30:55                          NaT   
422 2018-05-13 17:11:58                          NaT   
427 2018-01-30 22:34:14                          NaT   
429 2017-06-19 14:23:18                          NaT   

    order_delivered_customer_date order_estimated_delivery_date  \
50                            NaT                    2018-04-09   
330                           NaT                    2017-03-10   
422                           NaT                    2018-06-05   
427                           NaT                    2018-03-06   
429                           NaT                    2017-06-30   

     order_item_id                        product_id  \
50             NaN                               NaN   
330            NaN                               NaN   
422            NaN                               NaN   
427            1.0  25179680ecacb54a036e9cdd27459da3   
429            1.0  98354dc7140f034f3b55ec6fc45dd174   

                            seller_id shipping_limit_date  price  \
50                                NaN                 NaT    NaN   
330                               NaN                 NaT    NaN   
422                               NaN                 NaT    NaN   
427  b7cc6c5e001441ae8cdd5c69a480cbe4 2018-02-05 22:32:39  199.0   
429  7bcd7c5f8631701474db233ccf1c094b 2017-06-23 14:23:18   72.0   

     freight_value  
50             NaN  
330            NaN  
422            NaN  
427          83.25  
429          11.73

In [31]:
# Добавляю месяц на основе order_purchase_timestamp с помощью dt.strftime('%m')
df_1['month'] = df_1.order_purchase_timestamp.dt.strftime('%m').astype('int')

/opt/tljh/user/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [32]:
df_1

customer_id                customer_unique_id  \
50      f34a6e874087ec1f0e3dab9fdf659c5d  233896de79986082f1f479f1f85281cb   
330     5bfe800011656c0afb81db64519982db  0071f46a072a9ae25bbe4438b15efe9c   
422     b08064e24083fee8fbe8797902b07ecd  035f60af6e7d7f78470e9443be08d339   
427     3ba47b09546c2da91ef141efce2d5888  b8c37ee810b38c416022e1c9f51e740c   
429     cfffbb322333151f8df6398a8c3340db  0dc0731aafd5659410244092b0dc7d65   
...                                  ...                               ...   
112684  d9335c099e6943d88ac5f8a7cb418b66  22af178809fa6887af2cc4f7d7bdc0bc   
112685  d9335c099e6943d88ac5f8a7cb418b66  22af178809fa6887af2cc4f7d7bdc0bc   
112938  2034cc6d6d517007a9ba3e22ba643344  cb47d51801e7379766c5cd8ff623f37f   
113045  882b44646ab3f186597b1622a9443ff9  639b4096a4e5c96549a9bb495ca4871e   
113131  a10afcc34f66a704819e459a8f524ffa  6a65df7b3e532c05f3b312bb7e77ff33   

        customer_zip_code_prefix  customer_city customer_state  \
50                         38300      ituiutaba             MG   
330                        16011      aracatuba             SP   
422                        28495        aperibe             RJ   
427                        32140       contagem             MG   
429                         5093      sao paulo             SP   
...                          ...            ...            ...   
112684                      1218      sao paulo             SP   
112685                      1218      sao paulo             SP   
112938                     13052       campinas             SP   
113045                     13972        itapira             SP   
113131                     37160  campos gerais             MG   

                                order_id order_status  \
50      6e98de3a85c84ead6689189b825d35b5     canceled   
330     df8c077268f7f3baaac0892eb3143642  unavailable   
422     c609f82bcf7a90292a5940205ebd7e93  unavailable   
427     a402af38a74c44926247db438245d95c     canceled   
429     f63cd97f8f1db2eb266fb3bc2b4a95f3     canceled   
...                                  ...          ...   
112684  808c7c69c2778bdf4689eee0286e2bef     canceled   
112685  808c7c69c2778bdf4689eee0286e2bef     canceled   
112938  27d6012b882832dc9ddcf34304eeb77c     canceled   
113045  9b932dca249f9971b86dde2f2d7ad412  unavailable   
113131  8d171a848151e69e8f090fdbb50d4010     canceled   

       order_purchase_timestamp   order_approved_at  \
50          2018-03-15 10:07:02 2018-03-15 10:29:33   
330         2017-02-01 00:04:17 2017-02-01 00:30:55   
422         2018-05-13 16:45:55 2018-05-13 17:11:58   
427         2018-01-30 22:12:10 2018-01-30 22:34:14   
429         2017-06-19 14:07:04 2017-06-19 14:23:18   
...                         ...                 ...   
112684      2018-02-22 07:57:07 2018-02-22 08:10:27   
112685      2018-02-22 07:57:07 2018-02-22 08:10:27   
112938      2017-11-11 10:44:49 2017-11-11 10:55:28   
113045      2017-11-20 17:08:43 2017-11-20 17:26:54   
113131      2018-02-13 14:38:55 2018-02-15 04:07:13   

       order_delivered_carrier_date order_delivered_customer_date  \
50                              NaT                           NaT   
330                             NaT                           NaT   
422                             NaT                           NaT   
427                             NaT                           NaT   
429                             NaT                           NaT   
...                             ...                           ...   
112684                          NaT                           NaT   
112685                          NaT                           NaT   
112938                          NaT                           NaT   
113045                          NaT                           NaT   
113131          2018-02-16 00:19:33                           NaT   

       order_estimated_delivery_date  order_item_id  \
50                        2018-04-09            NaN   
330           

In [33]:
# Поссчитаем уникальное количество заказов, для каждого месяца и каждого выбранного статуса
order_id_uniq = df_1.groupby(['month', 'order_status'], as_index = False).agg({'order_id':'nunique'})

In [34]:
order_id_uniq.head()

month order_status  order_id
0      1     canceled        37
1      1  unavailable        58
2      2     canceled        90
3      2  unavailable        75
4      3     canceled        59

In [35]:
# Считаем среднее количество закзаов по месяцам для каждого статуса
order_id_mean = order_id_uniq.groupby('order_status', as_index = False).order_id.mean()

In [36]:
order_id_mean

order_status   order_id
0     canceled  52.083333
1  unavailable  50.750000

Задание 3. По каждому товару определить, в какой день недели товар чаще всего покупается.

1.Из order_purchase_timestamp получил день недели
2.Выделил день недели для каждого товара, которому соответсвует наибольшее количество продаж

In [37]:
# Получаю день недели с помощью dt.strftime('%A')
df['weekday'] = df.order_purchase_timestamp.dt.strftime('%A')

In [38]:
# Выделяю день недели для каждого товара, которому соответсвует наибольшее количество продаж с помощью max()
res_task_3 = df[['product_id', 'weekday']]\
                .groupby('product_id', as_index=False)\
                .agg({'weekday' : 'max'})

In [39]:
res_task_3

product_id    weekday
0      00066f42aeeb9f3007548bb9d3f33c38     Sunday
1      00088930e925c41fd95ebfe695fd2655    Tuesday
2      0009406fd7479715e4bef61dd91f2462   Thursday
3      000b8f95fcb9e0096488278317764d19  Wednesday
4      000d9be29b5207b54e86aa1b1ac54872    Tuesday
...                                 ...        ...
32946  fff6177642830a9a94a0f2cba5e476d1     Sunday
32947  fff81cc3158d2725c0655ab9ba0f712c     Monday
32948  fff9553ac224cec9d15d49f5a263411f     Friday
32949  fffdb2d0ec8d6a61f0a0a0db3f25b441    Tuesday
32950  fffe9eeff12fcbd74a2f2b007dde0c58  Wednesday

[32951 rows x 2 columns]

Задание 4. Сколько у каждого из пользователей в среднем покупок в неделю (по месяцам)? Не стоит забывать, что внутри месяца может быть не целое количество недель. Например, в ноябре 2021 года 4,28 недели. И внутри метрики это нужно учесть.

1.Добавил к датасету колонку с месяцем покупки
2.Рассчитал количество дней в датасете для каждого месяца
3.Выделил количество недель внутри месяца, поделив кол-во дней на 7
4.Рассчитал общее количество заказов для каждого пользователя внутри месяца
5.Поделил кол-во заказов для каждого пользователя внутри месяца на кол-во недель внутри месяца

In [40]:
res_task_4 = df[['customer_unique_id', 'order_id', 'order_status', 'order_purchase_timestamp']]\
                .query("order_status != 'unavailable' & order_status != 'canceled'")

In [41]:
res_task_4

customer_unique_id                          order_id  \
0       861eff4711a542e4b93843c6dd7febb0  00e7ee1b050b8499577073aeb2a297a1   
1       290c77bc529b7ac935b93aa66c333dc3  29150127e6685892b6eab3eec79f59c7   
2       060e732b5b29e8181a18229c7b0b2b5e  b2059ed67ce144a36e2aa97d2c9e9ad2   
3       259dac757896d24d7702b9acbbff3f3c  951670f92359f4fe4a63112aa7306eba   
4       345ecd01c38d18a9036ed96c73b8d066  6b7d50bd145f6fc7f33cebabd7e49d0f   
...                                  ...                               ...   
113420  1a29b476fee25c95fbafc67c5ac95cf8  6760e20addcf0121e9d58f2f1ff14298   
113421  d52a67c98be1cf6a5c84435bd38d095d  9ec0c8947d973db4f4e8dcf1fbfa8f1b   
113422  e9f50caf99f032f0bf3c55141f019d99  fed4434add09a6f332ea398efd656a5c   
113423  73c2643a0a458b49f58cea58833b192e  e31ec91cea1ecf97797787471f98a8c2   
113424  84732c5050c01db9b23e19ba39899398  28db69209a75e59f20ccbb5c36a20b90   

       order_status order_purchase_timestamp  
0         delivered      2017-05-16 15:05:35  
1         delivered      2018-01-12 20:48:24  
2         delivered      2018-05-19 16:07:45  
3         delivered      2018-03-13 16:06:38  
4         delivered      2018-07-29 09:51:30  
...             ...                      ...  
113420    delivered      2018-04-07 15:48:17  
113421    delivered      2018-04-04 08:20:22  
113422    delivered      2018-04-08 20:11:50  
113423    delivered      2017-11-03 21:08:33  
113424    delivered      2017-12-19 14:27:23  

[112109 rows x 4 columns]

In [42]:
# Добавляю к датасету колонку с месяцем покупки
res_task_4['year_month'] = res_task_4.order_purchase_timestamp.dt.strftime('%Y-%m')

In [43]:
res_task_4.head()

customer_unique_id                          order_id  \
0  861eff4711a542e4b93843c6dd7febb0  00e7ee1b050b8499577073aeb2a297a1   
1  290c77bc529b7ac935b93aa66c333dc3  29150127e6685892b6eab3eec79f59c7   
2  060e732b5b29e8181a18229c7b0b2b5e  b2059ed67ce144a36e2aa97d2c9e9ad2   
3  259dac757896d24d7702b9acbbff3f3c  951670f92359f4fe4a63112aa7306eba   
4  345ecd01c38d18a9036ed96c73b8d066  6b7d50bd145f6fc7f33cebabd7e49d0f   

  order_status order_purchase_timestamp year_month  
0    delivered      2017-05-16 15:05:35    2017-05  
1    delivered      2018-01-12 20:48:24    2018-01  
2    delivered      2018-05-19 16:07:45    2018-05  
3    delivered      2018-03-13 16:06:38    2018-03  
4    delivered      2018-07-29 09:51:30    2018-07

In [44]:
# Cчитаю количество дней в датасете для каждого месяца
res_task_4['days_in_month'] = res_task_4['order_purchase_timestamp'].dt.daysinmonth

In [45]:
# Выделяю количество недель внутри месяца, поделив кол-во дней на 7
res_task_4['weeks'] = res_task_4['days_in_month'] / 7

In [46]:
res_task_4

customer_unique_id                          order_id  \
0       861eff4711a542e4b93843c6dd7febb0  00e7ee1b050b8499577073aeb2a297a1   
1       290c77bc529b7ac935b93aa66c333dc3  29150127e6685892b6eab3eec79f59c7   
2       060e732b5b29e8181a18229c7b0b2b5e  b2059ed67ce144a36e2aa97d2c9e9ad2   
3       259dac757896d24d7702b9acbbff3f3c  951670f92359f4fe4a63112aa7306eba   
4       345ecd01c38d18a9036ed96c73b8d066  6b7d50bd145f6fc7f33cebabd7e49d0f   
...                                  ...                               ...   
113420  1a29b476fee25c95fbafc67c5ac95cf8  6760e20addcf0121e9d58f2f1ff14298   
113421  d52a67c98be1cf6a5c84435bd38d095d  9ec0c8947d973db4f4e8dcf1fbfa8f1b   
113422  e9f50caf99f032f0bf3c55141f019d99  fed4434add09a6f332ea398efd656a5c   
113423  73c2643a0a458b49f58cea58833b192e  e31ec91cea1ecf97797787471f98a8c2   
113424  84732c5050c01db9b23e19ba39899398  28db69209a75e59f20ccbb5c36a20b90   

       order_status order_purchase_timestamp year_month  days_in_month  \
0         delivered      2017-05-16 15:05:35    2017-05             31   
1         delivered      2018-01-12 20:48:24    2018-01             31   
2         delivered      2018-05-19 16:07:45    2018-05             31   
3         delivered      2018-03-13 16:06:38    2018-03             31   
4         delivered      2018-07-29 09:51:30    2018-07             31   
...             ...                      ...        ...            ...   
113420    delivered      2018-04-07 15:48:17    2018-04             30   
113421    delivered      2018-04-04 08:20:22    2018-04             30   
113422    delivered      2018-04-08 20:11:50    2018-04             30   
113423    delivered      2017-11-03 21:08:33    2017-11             30   
113424    delivered      2017-12-19 14:27:23    2017-12             31   

           weeks  
0       4.428571  
1       4.428571  
2       4.428571  
3       4.428571  
4       4.428571  
...          ...  
113420  4.285714  
113421  4.285714  
113422  4.285714  
113423  4.285714  
113424  4.428571  

[112109 rows x 7 columns]

In [47]:
# Считаю общее количество заказов для каждого пользователя внутри месяца
res_1_task_4 = res_task_4.groupby(['customer_unique_id', 'year_month', 'weeks'], as_index=False).agg({'order_id':'nunique'})

In [48]:
res_1_task_4

customer_unique_id year_month     weeks  order_id
0      0000366f3b9a7992bf8c76cfdf3221e2    2018-05  4.428571         1
1      0000b849f77a49e4a4ce2b2a4ca5be3f    2018-05  4.428571         1
2      0000f46a3911fa3c0805444483337064    2017-03  4.428571         1
3      0000f6ccb0745a6a4b88665a16c9f078    2017-10  4.428571         1
4      0004aac84e0df4da2b147fca70cf8255    2017-11  4.285714         1
...                                 ...        ...       ...       ...
96864  fffcf5a5ff07b0908bd4e2dbc735a684    2017-06  4.285714         1
96865  fffea47cd6d3cc0a88bd621562a9d061    2017-12  4.428571         1
96866  ffff371b4d645b6ecea244b27531430a    2017-02  4.000000         1
96867  ffff5962728ec6157033ef9805bacc48    2018-05  4.428571         1
96868  ffffd2657e2aad2907e67c3e9daecbeb    2017-05  4.428571         1

[96869 rows x 4 columns]

In [49]:
# Делю кол-во заказов для каждого пользователя внутри месяца на кол-во недель внутри месяца
res_1_task_4['orders_in_week'] = res_1_task_4.order_id / res_1_task_4.weeks

In [50]:
res_1_task_4.head(5)

customer_unique_id year_month     weeks  order_id  \
0  0000366f3b9a7992bf8c76cfdf3221e2    2018-05  4.428571         1   
1  0000b849f77a49e4a4ce2b2a4ca5be3f    2018-05  4.428571         1   
2  0000f46a3911fa3c0805444483337064    2017-03  4.428571         1   
3  0000f6ccb0745a6a4b88665a16c9f078    2017-10  4.428571         1   
4  0004aac84e0df4da2b147fca70cf8255    2017-11  4.285714         1   

   orders_in_week  
0        0.225806  
1        0.225806  
2        0.225806  
3        0.225806  
4        0.233333

In [51]:
res_1_task_4.sort_values('orders_in_week')

customer_unique_id year_month     weeks  order_id  \
0      0000366f3b9a7992bf8c76cfdf3221e2    2018-05  4.428571         1   
56542  9585898e80f9de26178eb04cd600b6f5    2017-08  4.428571         1   
56543  958843e128a10a8ae35155b6fc57fec3    2018-03  4.428571         1   
56546  958a09f1a17bfaa6ca721b5aa01f59f6    2018-05  4.428571         1   
56547  958ac7760e5484df1496ba5833ab6117    2018-08  4.428571         1   
...                                 ...        ...       ...       ...   
14312  25a560b9a6006157838aab1bdbd68624    2017-04  4.285714         4   
23654  3e43e6105506432c953e165fb2acf44c    2018-02  4.000000         4   
61451  a239b8e2fbce33780f1f1912e2ee5275    2017-02  4.000000         4   
68556  b4e4f24de1e8725b74e4a1f4975116ed    2018-02  4.000000         4   
7232   12f5d6e1cbf93dafd9dcc19095df0b3d    2017-01  4.428571         6   

       orders_in_week  
0            0.225806  
56542        0.225806  
56543        0.225806  
56546        0.225806  
56547        0.225806  
...               ...  
14312        0.933333  
23654        1.000000  
61451        1.000000  
68556        1.000000  
7232         1.354839  

[96869 rows x 5 columns]

"""
Задание 5. Используя pandas, проведи когортный анализ пользователей. В период с января по декабрь выяви когорту с самым высоким retention на 3й месяц. 

1.Когорты строил на время создания заказа order_purchase_timestamp
2.Создал новый столбец year_month, который является годом и месяцем, когда произошел первый заказ клиента
3.Агрегировал клиентов, покупки и суммы покупок по каждой year_month с помощью min()
4.Создал сводную таблицу с помощью pivot_table
5.Выявил когорту с самым высоким retention на 3й месяц
"""

In [52]:
res_task_5 = df[['customer_unique_id', 'order_purchase_timestamp']]

In [53]:
res_task_5

customer_unique_id order_purchase_timestamp
0       861eff4711a542e4b93843c6dd7febb0      2017-05-16 15:05:35
1       290c77bc529b7ac935b93aa66c333dc3      2018-01-12 20:48:24
2       060e732b5b29e8181a18229c7b0b2b5e      2018-05-19 16:07:45
3       259dac757896d24d7702b9acbbff3f3c      2018-03-13 16:06:38
4       345ecd01c38d18a9036ed96c73b8d066      2018-07-29 09:51:30
...                                  ...                      ...
113420  1a29b476fee25c95fbafc67c5ac95cf8      2018-04-07 15:48:17
113421  d52a67c98be1cf6a5c84435bd38d095d      2018-04-04 08:20:22
113422  e9f50caf99f032f0bf3c55141f019d99      2018-04-08 20:11:50
113423  73c2643a0a458b49f58cea58833b192e      2017-11-03 21:08:33
113424  84732c5050c01db9b23e19ba39899398      2017-12-19 14:27:23

[113425 rows x 2 columns]

In [54]:
# Поскольку я делаю ежемесячные когорты, я буду смотреть на общее ежемесячное поведение наших клиентов.
# Создадим новый столбец year_month, который является годом и месяцем, когда произошел первый заказ клиента.
res_task_5['year_month'] = res_task_5.order_purchase_timestamp.dt.strftime('%Y-%m')

/opt/tljh/user/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [55]:
res_task_5

customer_unique_id order_purchase_timestamp year_month
0       861eff4711a542e4b93843c6dd7febb0      2017-05-16 15:05:35    2017-05
1       290c77bc529b7ac935b93aa66c333dc3      2018-01-12 20:48:24    2018-01
2       060e732b5b29e8181a18229c7b0b2b5e      2018-05-19 16:07:45    2018-05
3       259dac757896d24d7702b9acbbff3f3c      2018-03-13 16:06:38    2018-03
4       345ecd01c38d18a9036ed96c73b8d066      2018-07-29 09:51:30    2018-07
...                                  ...                      ...        ...
113420  1a29b476fee25c95fbafc67c5ac95cf8      2018-04-07 15:48:17    2018-04
113421  d52a67c98be1cf6a5c84435bd38d095d      2018-04-04 08:20:22    2018-04
113422  e9f50caf99f032f0bf3c55141f019d99      2018-04-08 20:11:50    2018-04
113423  73c2643a0a458b49f58cea58833b192e      2017-11-03 21:08:33    2017-11
113424  84732c5050c01db9b23e19ba39899398      2017-12-19 14:27:23    2017-12

[113425 rows x 3 columns]

In [56]:
# Теперь, нам необходимо агрегировать клиентов, покупки и суммы покупок по каждой year_month
cohorts = res_task_5.groupby(['customer_unique_id','order_purchase_timestamp'], as_index=False).agg({'year_month': 'min'}).rename(columns={'year_month': 'cohortgroup'})

In [57]:
cohorts['ret_month'] = cohorts.groupby('customer_unique_id', as_index=False).cohortgroup.cumcount()

In [58]:
cohorts['retention'] = 1

In [59]:
cohorts

customer_unique_id order_purchase_timestamp cohortgroup  \
0      0000366f3b9a7992bf8c76cfdf3221e2      2018-05-10 10:56:27     2018-05   
1      0000b849f77a49e4a4ce2b2a4ca5be3f      2018-05-07 11:11:27     2018-05   
2      0000f46a3911fa3c0805444483337064      2017-03-10 21:05:03     2017-03   
3      0000f6ccb0745a6a4b88665a16c9f078      2017-10-12 20:29:41     2017-10   
4      0004aac84e0df4da2b147fca70cf8255      2017-11-14 19:45:42     2017-11   
...                                 ...                      ...         ...   
99144  fffcf5a5ff07b0908bd4e2dbc735a684      2017-06-08 21:00:36     2017-06   
99145  fffea47cd6d3cc0a88bd621562a9d061      2017-12-10 20:07:56     2017-12   
99146  ffff371b4d645b6ecea244b27531430a      2017-02-07 15:49:16     2017-02   
99147  ffff5962728ec6157033ef9805bacc48      2018-05-02 15:17:41     2018-05   
99148  ffffd2657e2aad2907e67c3e9daecbeb      2017-05-02 20:18:45     2017-05   

       ret_month  retention  
0              0          1  
1              0          1  
2              0          1  
3              0          1  
4              0          1  
...          ...        ...  
99144          0          1  
99145          0          1  
99146          0          1  
99147          0          1  
99148          0          1  

[99149 rows x 5 columns]

In [60]:
# Создаю сводную таблицу с помощью pivot_table
cohorts = pd.pivot_table(cohorts, index='cohortgroup', columns='ret_month', values='retention', aggfunc=np.sum, fill_value=0)

In [80]:
cohorts

ret_month      0    1   2   3   4   5   6   7   8   9   10  11  12  13  14  \
cohortgroup                                                                  
2016-09         4    0   0   0   0   0   0   0   0   0   0   0   0   0   0   
2016-10       321    3   0   0   0   0   0   0   0   0   0   0   0   0   0   
2016-12         1    0   0   0   0   0   0   0   0   0   0   0   0   0   0   
2017-01       764   21   4   2   1   1   0   0   0   0   0   0   0   0   0   
2017-02      1752   21   2   1   0   0   0   0   0   0   0   0   0   0   0   
2017-03      2636   39   1   0   0   0   0   0   0   0   0   0   0   0   0   
2017-04      2352   41   3   1   0   0   0   0   0   0   0   0   0   0   0   
2017-05      3596   84   3   0   0   0   0   0   0   0   0   0   0   0   0   
2017-06      3139   89   8   0   0   0   0   0   0   0   0   0   0   0   0   
2017-07      3894  104   7   2   1   0   0   0   0   0   0   0   0   0   0   
2017-08      4184  114   7   1   0   1   0   0   0   0   0   0   0   0   0   
2017-09      4130  125  12   1   0   0   1   0   0   0   0   0   0   0   0   
2017-10      4470  132  13   2   0   0   0   1   1   0   0   0   0   0   0   
2017-11      7304  193  16   4   1   0   0   0   0   1   0   0   0   0   0   
2017-12      5487  153   9   4   2   0   0   0   0   0   0   0   0   0   0   
2018-01      7025  203  14   2   3   3   0   0   0   0   0   0   0   0   0   
2018-02      6451  224  14   3   2   1   2   1   1   0   0   0   0   0   0   
2018-03      6965  200  18   3   2   0   0   0   0   0   0   0   0   0   0   
2018-04      6711  191  20   3   2   1   0   0   0   0   0   0   0   0   0   
2018-05      6622  219  20   4   0   2   0   0   0   0   1   1   0   0   0   
2018-06      5940  190  12   8   2   0   2   0   0   0   0   0   0   0   0   
2018-07      6071  180  13   1   0   0   0   0   0   0   0   0   1   1   0   
2018-08      6271  204  22   5   1   1   0   0   0   0   0   0   0   0   1   
2018-09         5    8   3   0   0   0   0   0   0   0   0   0   0   0   0   
2018-10         1    2   1   0   0   0   0   0   0   0   0   0   0   0   0   

ret_month    15  16  
cohortgroup          
2016-09       0   0  
2016-10       0   0  
2016-12       0   0  
2017-01       0   0  
2017-02       0   0  
2017-03       0   0  
2017-04       0   0  
2017-05       0   0  
2017-06       0   0  
2017-07       0   0  
2017-08       0   0  
2017-09       0   0  
2017-10       0   0  
2017-11       0   0  
2017-12       0   0  
2018-01       0   0  
2018-02       0   0  
2018-03       0   0  
2018-04       0   0  
2018-05       0   0  
2018-06       0   0  
2018-07       0   0  
2018-08       1   1  
2018-09       0   0  
2018-10       0   0

In [61]:
# В период с января по декабрь  с самым высоким retention на 3й месяц. 
cohorts[3].max()

8

In [62]:
# Это соответсвует когорте '2018-06'

"""
Задание 6. Часто для качественного анализа аудитории использую подходы, основанные на сегментации. Используя python, построй RFM-сегментацию пользователей, чтобы качественно оценить свою аудиторию. В кластеризации можешь выбрать следующие метрики: R - время от последней покупки пользователя до текущей даты, F - суммарное количество покупок у пользователя за всё время, M - сумма покупок за всё время. Подробно опиши, как ты создавал кластеры. Для каждого RFM-сегмента построй границы метрик recency, frequency и monetary для интерпретации этих кластеров. Пример такого описания: RFM-сегмент 132 (recency=1, frequency=3, monetary=2) имеет границы метрик recency от 130 до 500 дней, frequency от 2 до 5 заказов в неделю, monetary от 1780 до 3560 рублей в неделю.


1.Выбрать методику расчета RFM-кластеров
2.Я выбрал простейшую для 1 варианта:
    R - разница между текущей датой и датой последнего заказа (в днях с последнего заказа)
    F - количество заказов за все время
    M - общая сумма покупок за все время
3.Кластеризация: разбиение по квартилям этих метрик
4.Рассчитал время от каждого заказа до максимально возможного времени внутри датасета
5.Рассчитал стоимость заказа для каждого заказа в данных
6.Для каждого пользователя вычислил:
    срок в днях от его последнего заказа
    суммарное количество заказов
    общую сумму заказов
7.Выставил оценки, разбивая на диапазон от 1 до 5

"""

In [63]:
# Я буду опираться на order_purchase_timestamp —  время создания заказа, как и в предыдущих заданиях. 
df.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  \
0                     14409                 franca             SP   
1                      9790  sao bernardo do campo             SP   
2                      1151              sao paulo             SP   
3                      8775        mogi das cruzes             SP   
4                     13056               campinas             SP   

                           order_id order_status order_purchase_timestamp  \
0  00e7ee1b050b8499577073aeb2a297a1    delivered      2017-05-16 15:05:35   
1  29150127e6685892b6eab3eec79f59c7    delivered      2018-01-12 20:48:24   
2  b2059ed67ce144a36e2aa97d2c9e9ad2    delivered      2018-05-19 16:07:45   
3  951670f92359f4fe4a63112aa7306eba    delivered      2018-03-13 16:06:38   
4  6b7d50bd145f6fc7f33cebabd7e49d0f    delivered      2018-07-29 09:51:30   

    order_approved_at order_delivered_carrier_date  \
0 2017-05-16 15:22:12          2017-05-23 10:47:57   
1 2018-01-12 20:58:32          2018-01-15 17:14:59   
2 2018-05-20 16:19:10          2018-06-11 14:31:00   
3 2018-03-13 17:29:19          2018-03-27 23:22:42   
4 2018-07-29 10:10:09          2018-07-30 15:16:00   

  order_delivered_customer_date order_estimated_delivery_date  order_item_id  \
0           2017-05-25 10:35:35                    2017-06-05            1.0   
1           2018-01-29 12:41:19                    2018-02-06            1.0   
2           2018-06-14 17:58:51                    2018-06-13            1.0   
3           2018-03-28 16:04:25                    2018-04-10            1.0   
4           2018-08-09 20:55:48                    2018-08-15            1.0   

                         product_id                         seller_id  \
0  a9516a079e37a9c9c36b9b78b10169e8  7c67e1448b00f6e969d365cea6b010ab   
1  4aa6014eceb682077f9dc4bffebc05b0  b8bc237ba3788b23da09c0f1f3a3288c   
2  bd07b66896d6f1494f5b86251848ced7  7c67e1448b00f6e969d365cea6b010ab   
3  a5647c44af977b148e0a3a4751a09e2e  7c67e1448b00f6e969d365cea6b010ab   
4  9391a573abe00141c56e38d84d7d5b3b  4a3ca9315b744ce9f8e9374361493884   

  shipping_limit_date   price  freight_value   weekday  
0 2017-05-22 15:22:12  124.99          21.88   Tuesday  
1 2018-01-18 20:58:32  289.00          46.48    Friday  
2 2018-06-05 16:19:10  139.94          17.79  Saturday  
3 2018-03-27 16:31:16  149.94          23.36   Tuesday  
4 2018-07-31 10:10:09  230.00          22.25    Sunday

In [64]:
# Вместо текущей даты будем использовать max+1. 
from datetime import timedelta

In [65]:
last_date = df['order_purchase_timestamp'].max() + timedelta(days=1)

In [66]:
# Для создания recency, frequency и monetary я группирую df по customer_id.
rfm = df.reset_index().groupby('customer_id')\
        .agg({'order_purchase_timestamp': lambda x: (last_date - x.max()).days,
             'order_id': 'count',
             'price': 'sum'})

In [67]:
rfm.rename(columns={'order_purchase_timestamp': 'recency',
                   'order_id': 'frequency',
                   'price': 'monetary'}, inplace=True)

In [68]:
rfm

recency  frequency  monetary
customer_id                                                   
00012a2ce6f8dcda20d059ce98491703      338          1     89.80
000161a058600d5901f007fab4c27140      459          1     54.90
0001fd6190edaaf884bcaf3d49edf079      597          1    179.99
0002414f95344307404f0ace7a26f1d5      428          1    149.90
000379cdec625522490c315e70c7a9fb      199          1     93.00
...                                   ...        ...       ...
fffecc9f79fd8c764f843e9951b11341      203          1     54.90
fffeda5b6d849fbd39689bb92087f431      149          1     47.90
ffff42319e9b2d713724ae527742af25      127          1    199.90
ffffa3172527f765de70084a7e53aae8      411          2     21.80
ffffe8b65bbe3087b653a978c870db99      384          1      0.00

[99441 rows x 3 columns]

In [69]:
# Теперь нам нужно выставить оценки. 
# Мы будем разбивать на диапазон от 1 до 4. 
#Чем шире диапазон, тем более точные наши группы, но в тоже время работать с большим количествов комбинаций сложнее.
quantiles = rfm.quantile(q=[0.2, 0.4, 0.6, 0.8])

In [70]:
quantiles

recency  frequency  monetary
0.2    146.0        1.0     38.17
0.4    230.0        1.0     66.00
0.6    321.0        1.0    105.00
0.8    437.0        1.0    174.99

In [71]:
quantiles = quantiles.to_dict()

In [72]:
quantiles

{'recency': {0.2: 146.0, 0.4: 230.0, 0.6: 321.0, 0.8: 437.0},
 'frequency': {0.2: 1.0, 0.4: 1.0, 0.6: 1.0, 0.8: 1.0},
 'monetary': {0.2: 38.17, 0.4: 66.0, 0.6: 105.0, 0.8: 174.99}}

In [73]:
def get_segment(x,p,d):
    if x <= d[p][0.2]:
        return 1
    elif x <= d[p][0.4]:
        return 2
    elif x <= d[p][0.6]:
        return 3
    elif x <= d[p][0.8]:
        return 4
    else:
        return 5

In [74]:
rfm['R'] = rfm['recency'].apply(get_segment, args=('recency', quantiles,))

In [75]:
rfm['F'] = rfm['frequency'].apply(get_segment, args=('frequency', quantiles,))

In [76]:
rfm['M'] = rfm['monetary'].apply(get_segment, args=('monetary', quantiles,))

In [77]:
rfm['RFM_Score'] = rfm.R.map(str) + rfm.F.map(str) + rfm.M.map(str)

In [78]:
rfm

recency  frequency  monetary  R  F  M  \
customer_id                                                               
00012a2ce6f8dcda20d059ce98491703      338          1     89.80  4  1  3   
000161a058600d5901f007fab4c27140      459          1     54.90  5  1  2   
0001fd6190edaaf884bcaf3d49edf079      597          1    179.99  5  1  5   
0002414f95344307404f0ace7a26f1d5      428          1    149.90  4  1  4   
000379cdec625522490c315e70c7a9fb      199          1     93.00  2  1  3   
...                                   ...        ...       ... .. .. ..   
fffecc9f79fd8c764f843e9951b11341      203          1     54.90  2  1  2   
fffeda5b6d849fbd39689bb92087f431      149          1     47.90  2  1  2   
ffff42319e9b2d713724ae527742af25      127          1    199.90  1  1  5   
ffffa3172527f765de70084a7e53aae8      411          2     21.80  4  5  1   
ffffe8b65bbe3087b653a978c870db99      384          1      0.00  4  1  1   

                                 RFM_Score  
customer_id                                 
00012a2ce6f8dcda20d059ce98491703       413  
000161a058600d5901f007fab4c27140       512  
0001fd6190edaaf884bcaf3d49edf079       515  
0002414f95344307404f0ace7a26f1d5       414  
000379cdec625522490c315e70c7a9fb       213  
...                                    ...  
fffecc9f79fd8c764f843e9951b11341       212  
fffeda5b6d849fbd39689bb92087f431       212  
ffff42319e9b2d713724ae527742af25       115  
ffffa3172527f765de70084a7e53aae8       451  
ffffe8b65bbe3087b653a978c870db99       411  

[99441 rows x 7 columns]

In [79]:
# Оценки RFM дают нам 5^3 = 125 сегментов. С которым нелегко работать. Надо выбрать определенные сегменты удовлеворяющие требованиям поставленных задач. Например: удержание клиентов, повышение дохода, повышение лояльность клиентов.

In [81]:
df

customer_id                customer_unique_id  \
0       06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1       18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2       4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3       b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4       4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                  ...                               ...   
113420  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
113421  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
113422  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
113423  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
113424  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

        customer_zip_code_prefix          customer_city customer_state  \
0                          14409                 franca             SP   
1                           9790  sao bernardo do campo             SP   
2                           1151              sao paulo             SP   
3                           8775        mogi das cruzes             SP   
4                          13056               campinas             SP   
...                          ...                    ...            ...   
113420                      3937              sao paulo             SP   
113421                      6764        taboao da serra             SP   
113422                     60115              fortaleza             CE   
113423                     92120                 canoas             RS   
113424                      6703                  cotia             SP   

                                order_id order_status  \
0       00e7ee1b050b8499577073aeb2a297a1    delivered   
1       29150127e6685892b6eab3eec79f59c7    delivered   
2       b2059ed67ce144a36e2aa97d2c9e9ad2    delivered   
3       951670f92359f4fe4a63112aa7306eba    delivered   
4       6b7d50bd145f6fc7f33cebabd7e49d0f    delivered   
...                                  ...          ...   
113420  6760e20addcf0121e9d58f2f1ff14298    delivered   
113421  9ec0c8947d973db4f4e8dcf1fbfa8f1b    delivered   
113422  fed4434add09a6f332ea398efd656a5c    delivered   
113423  e31ec91cea1ecf97797787471f98a8c2    delivered   
113424  28db69209a75e59f20ccbb5c36a20b90    delivered   

       order_purchase_timestamp   order_approved_at  \
0           2017-05-16 15:05:35 2017-05-16 15:22:12   
1           2018-01-12 20:48:24 2018-01-12 20:58:32   
2           2018-05-19 16:07:45 2018-05-20 16:19:10   
3           2018-03-13 16:06:38 2018-03-13 17:29:19   
4           2018-07-29 09:51:30 2018-07-29 10:10:09   
...                         ...                 ...   
113420      2018-04-07 15:48:17 2018-04-07 16:08:45   
113421      2018-04-04 08:20:22 2018-04-04 08:35:12   
113422      2018-04-08 20:11:50 2018-04-08 20:30:03   
113423      2017-11-03 21:08:33 2017-11-03 21:31:20   
113424      2017-12-19 14:27:23 2017-12-19 18:50:39   

       order_delivered_carrier_date order_delivered_customer_date  \
0               2017-05-23 10:47:57           2017-05-25 10:35:35   
1               2018-01-15 17:14:59           2018-01-29 12:41:19   
2               2018-06-11 14:31:00           2018-06-14 17:58:51   
3               2018-03-27 23:22:42           2018-03-28 16:04:25   
4               2018-07-30 15:16:00           2018-08-09 20:55:48   
...                             ...                           ...   
113420          2018-04-11 02:08:36           2018-04-13 20:06:37   
113421          2018-04-05 18:42:35           2018-04-11 18:54:45   
113422          2018-04-09 17:52:17           2018-05-09 19:03:15   
113423          2017-11-06 18:24:41           2017-11-16 19:58:39   
113424          2017-12-21 19:17:21           2017-12-26 18:42:36   

       order_estimated_deliv

customer_id                customer_unique_id  \
109431  ffffe8b65bbe3087b653a978c870db99  736e6bfa0510aa5b878881a226a5fd89   

        customer_zip_code_prefix customer_city customer_state  \
109431                      6172        osasco             SP   

                                order_id order_status  \
109431  2e935fa1d39497aa0ec3f1107fbfb5b8  unavailable   

       order_purchase_timestamp   order_approved_at  \
109431      2017-09-29 14:07:03 2017-10-03 06:34:13   

       order_delivered_carrier_date order_delivered_customer_date  \
109431                          NaT                           NaT   

       order_estimated_delivery_date  order_item_id product_id seller_id  \
109431                    2017-10-24            NaN        NaN       NaN   

       shipping_limit_date  price  freight_value weekday  
109431                 NaT    NaN            NaN  Friday